File: harmony/pushy/src/main/ets/NativeUpdateResult.ts
/** A native round snapshot, not the current state of the running RN instance. */
export interface NativeUpdateResult {
  /** skipped, noUpdate, downloaded, failed, or cancelled. */
  status: string;
  reason: string;
  hash: string;
  /** The version was selected for the next launch; no reload is performed. */
  activated: boolean;
}

export function nativeUpdateResult(
  status: string,
  reason: string = '',
  hash: string = '',
  activated: boolean = false,
): NativeUpdateResult {
  return { status, reason, hash, activated };
}

/** Internal gate shared by the delayed check and the native host API. */
export class NativeUpdateRound {
  private task: Promise<NativeUpdateResult> | undefined;

  run(operation: () => Promise<NativeUpdateResult>): Promise<NativeUpdateResult> {
    if (this.task !== undefined) {
      return this.task;
    }
    // Defer invocation until the promise is stored, including reentrant callers.
    this.task = Promise.resolve().then(operation);
    return this.task;
  }
}

File: android/src/main/java/cn/reactnative/modules/update/NativeUpdateResult.java
package cn.reactnative.modules.update;

/**
 * Snapshot of a native check-and-update round. A downloaded result does not
 * mean that the running React Native instance has been reloaded.
 */
public final class NativeUpdateResult {
    public static final String SKIPPED = "skipped";
    public static final String NO_UPDATE = "noUpdate";
    public static final String DOWNLOADED = "downloaded";
    public static final String FAILED = "failed";
    public static final String CANCELLED = "cancelled";

    private final String status;
    private final String reason;
    private final String hash;
    private final boolean activated;

    private NativeUpdateResult(String status, String reason, String hash, boolean activated) {
        this.status = status;
        this.reason = reason;
        this.hash = hash;
        this.activated = activated;
    }

    static NativeUpdateResult of(String status, String reason) {
        return new NativeUpdateResult(status, reason, "", false);
    }

    static NativeUpdateResult downloaded(String hash, boolean activated) {
        return new NativeUpdateResult(DOWNLOADED, "", hash, activated);
    }

    public String getStatus() {
        return status;
    }

    public String getReason() {
        return reason;
    }

    /** Installed update hash, or an empty string when this round installed nothing. */
    public String getHash() {
        return hash;
    }

    /** Whether this round selected the downloaded version for the next launch. */
    public boolean isActivated() {
        return activated;
    }
}
